internal/controller/clusterinstance/template_engine.go

```diff
@@ -20,50 +20,54 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"strings"
 	"text/template"
 	"unicode"
 
+	sprig "github.com/go-task/slim-sprig"
 	"go.uber.org/zap"
 	"gopkg.in/yaml.v3"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	k8syaml "sigs.k8s.io/yaml"
 
 	"github.com/stolostron/siteconfig/api/v1alpha1"
+	"github.com/stolostron/siteconfig/internal/controller/common"
+	"github.com/stolostron/siteconfig/internal/templates/engine"
 )
 
-const (
-	WaveAnnotation        = v1alpha1.Group + "/sync-wave"
-	DefaultWaveAnnotation = "0"
-	OwnedByLabel          = v1alpha1.Group + "/owned-by"
-)
-
-type TemplateEngine struct {
-	log *zap.Logger
-}
+// CITemplateEngine is a concrete implementation of the TemplateEngine interface for the
+// ClusterInstance resource
+type CITemplateEngine struct{}
 
-// SetLogger is to be used to log with ClusterInstance fields
-func (te *TemplateEngine) SetLogger(log1 *zap.Logger) {
-	te.log = log1
+// NewCITemplateEngine creates a new instance of ConcreteTemplateEngine
+func NewCITemplateEngine() engine.TemplateEngine {
+	return &CITemplateEngine{}
 }
 
-func NewTemplateEngine(defaultLogger *zap.Logger) *TemplateEngine {
-	return &TemplateEngine{log: defaultLogger}
-}
-
-func (te *TemplateEngine) ProcessTemplates(
+func (te *CITemplateEngine) ProcessTemplates(
 	ctx context.Context,
 	c client.Client,
-	clusterInstance v1alpha1.ClusterInstance,
-) (RenderedObjectCollection, error) {
+	log *zap.Logger,
+	object interface{},
+) (engine.RenderedObjectCollection, error) {
+
+	log = log.Named("ProcessTemplates")
+
+	var renderedObjects engine.RenderedObjectCollection
 
-	log := te.log.Named("ProcessTemplates")
+	// Type assertion to handle specific object types
+	clusterInstance, ok := object.(v1alpha1.ClusterInstance)
```

Comment:
Can this ever be something else? Why not pass the concrete type rather than `interface{}`?

Reply:
The TemplateEngine interface is future-proof designed such that it can accept the processing of an arbitrary CR.